eclipse-che/che-plugin-registry
# Run VSCode extension's tests on top of che-theia

### What does this PR do?
This PR contains an action **VSCode Extension's Tests** to run tests from VSCode extension on top of Che-Theia if it was changed in `che-theia-plugins.yaml`.

VSCode Extension's Tests workflow:
https://github.com/svor/che-plugin-registry/actions/workflows/vscode-extension-tests-check.yaml
![screenshot-github com-2021 03 01-14_50_59](https://user-images.githubusercontent.com/1271546/109499467-9d278880-7a9d-11eb-859a-340760290f41.png)
1. The first step is **Checkout**. It clones the repository and makes checkout to the working branch.

2. Step 2 is **Get changes**. This step executes `./ci/get-changed-plugin.sh`. In this script we check what exactly was changed in `che-theia-plugins.yaml`. If this file doesn't contain changes on lines with key `revision` - it means that none extension is going to be updated or added and in this way we set `EXTENSION_REPO` environment variable as empty and all next action's steps will be skipped (action should be finished). In another way, when some value of revision was changed, we find all information about the repository of changed extension (url and a revision).

3. Step 3 is **Build image**. It builds `che-plugin-registry` image with all updates. It is needed to run the workspace with updated/added plugin.

4. Step 4 is **Start minikube**. Needs to setup minikube

5. Step 5 is **Load image**. It loads che-plugin registry image into docker registry

6. Step 6 is **Deploy Eclipse Che**. This step uses `che-incubator/che-deploy-action@next` action to deploy Che with custom che-plugin-registry image into minikube.

7. Step 7 is **Run VSCode extension's tests**. This step executes `./ci/run-extension-tests.sh`. The script has a few steps:
	-  First of all we need to install all required dependencies into the runner (selenium, geckodriver)
	-  By using `url` and `revision` from step #2 we clone extension repository into the` /tmp/projects/` folder and do checkout to the branch
	-  Compile extension (`yarn install` and `yarn compile`  --- I'm not sure if it will work for all extension)
	-  Next step is to prepare a devfile to create tested workspace. A template of this devfile is in `.ci/templates/extension-tests-devfile.yaml` and it looks like 

```yaml
apiVersion: 1.0.0
metadata:
  name: extension-test-sample
components:
  - mountSources: true
    memoryLimit: 3Gi
    type: dockerimage
    image: 'quay.io/eclipse/che-theia-dev:next'
    alias: che-dev
  - type: chePlugin
    reference: 'https://raw.githubusercontent.com/svor/che-vscode-extension-tests/main/meta.yaml'
  - type: chePlugin
    id: @
  - id: eclipse/che-theia/next
    memoryLimit: 1024Mi
    type: cheEditor
```

- When the devile is ready we can start the workspace. By using chectl we start the workspace and check when its status will be Running. When the workspce is ready, we copy extension's sources into /projects in theia-ide container. After that by using python .ci/tests-runner.py we run selenium script to open the workspace and run extension's tests.
- The last step is to check /projects/test.log file which should contain the result of test execution.


So, this step builds id of tested extension by analysing `che-theia-plugins.yaml`. If the changed plugin contains id we just take it, if not we build it from extension's `package.jon` as `publisher/name/latest`. When we have the id we need to replace @ with it in the devfile.



**How are tests running.**

The workspace contains `che-vscode-extension-tests` plug-in (it locates in local repository https://github.com/svor/che-vscode-extension-tests) it is included in the devfile:
```yaml
  - type: chePlugin
    reference: 'https://raw.githubusercontent.com/svor/che-vscode-extension-tests/main/meta.yaml'
```
When the workspace was started, this plugin automatically finds all *.test.js files in `/projects` folder inside the workspace and run them one by one via mocha framework.

For now tests are failed because each test needs to execute vscode.extensions.getExtension(id) to activate an extension and it gets `Cannot read property 'getExtension' of undefined`. There is an issue to investigate this problem: https://github.com/eclipse/che/issues/19172


### Screenshot/screencast of this PR
Some details of execution is here https://github.com/svor/che-plugin-registry/runs/1979560988?check_suite_focus=true


### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/18219

### How to test this PR?



### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.
